core/devices/machine_topology.cpp

```diff
@@ -0,0 +1,263 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2021, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#include <atomic>
+#include <memory>
+#include <mutex>
+
+
+#include <ginkgo/core/base/machine_topology.hpp>
+
+
+namespace gko {
+
+
+namespace detail {
+
+
+class topo_bitmap {
+public:
+#if GKO_HAVE_HWLOC
+    using bitmap_type =
+        std::remove_pointer<decltype(hwloc_bitmap_alloc())>::type;
```

Comment:
I think it is somewhat safe to assume that `hwloc_bitmap_alloc()` returns a pointer (since `alloc` usually returns a pointer). However, in this context, it is not necessary to know that it is a pointer, so removing it would be even more safe:
```suggestion
    using bitmap_type =
        decltype(hwloc_bitmap_alloc())::type;
```
Alternatively, you could obviously use `hwloc_bitmap_t` directly.
When I wrote this part as a suggestion, I had not idea what it actually returned, and I needed to remove the pointer in order to work with `unique_ptr`.

Reply:
I have updated this now.